Geometry simplifier (streaming)

craigtaverner opened the pull request:
There are many opportunities to enable geometry simplification in Elasticsearch, both as an explicit feature available to users, and as an internal optimization technique for reducing memory consumption for complex geometries. For the latter case, it can even be considered a bug fix. This PR provides support for constraining Line and LinearRing sizes to a fixed number of points, and thereby a fixed amount of memory usage.

Consider, for example, the geo_line aggregation. This is similar to the top-10 aggregation, but allows the top-10k (ten thousand) points to be aggregated. This is not only a lot of memory, but can still cause unwanted line truncation for very large geometries. Line simplification is a solution to this. It is likely that a much smaller limit than 10k would suffice, while at the same time not truncating the geometry at all, so we fix a bug (truncation) while improving memory usage (pull limit from 10k down to perhaps just 1k).

This PR provides two APIs:

* Streaming:
  * By using the `simplifier.consume(x, y)` method on a stream of points, the total memory used is limited to a linear function of `k`, the total number of points to retain. This algorithm is at its heart based on the [Visvalingam–Whyatt algorithm](https://en.wikipedia.org/wiki/Visvalingam%E2%80%93Whyatt_algorithm), with concepts from https://bost.ocks.org/mike/simplify/ and in particular the detailed streaming discussions in the paper at https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.106.7132&rep=rep1&type=pdf
* Full-geometry:
  * Simplifying full geometries using the `simplifier.simplify(geometry)` method can work with most geometry types, even `GeometryCollection`, but:
    * Some geometries do not get simplified because it makes no sense to: Point, Circle, Rectangle
    * The `maxPoints` parameter is used as is to apply to the main component (shell for polygons, largest geometry for multi-polygons and geometry collections), and all other sub-components (holes in polygons, etc.) are simplified to a scaled down version of the `maxPoints`, scaled by the relative size of the sub-component to the main component.
  * The simplification itself is done on each Line and LinearRing component using the same streaming algorithm above. Since we use the [Visvalingam–Whyatt algorithm](https://en.wikipedia.org/wiki/Visvalingam%E2%80%93Whyatt_algorithm), this works is applicable to both streaming and full-geometry simplification with the same essential result, but better control over memory than normal full-geometry simplifiers.

The basic algorithm for simplification on a stream of points requires maintaining two data structures:

* an array of all currently simplified points (implicitly ordered in stream order)
* a priority queue of all but the two end points with an estimated error on each that expresses the cost of removing that point from the line

Then running the algorithm on each incoming point:

* If the incoming point is the third point or later:
  * Evaluate the simplification error of the previous point between its previous and the new one being added
  * Add that previous point and its error to the priority queue (note how this leads to the priority queue having two fewer points than the sorted array)
* If the total number of points in the sorted array has reached the maximum point count:
  * Determine which point in the priority queue has the lowest error (least impact to the geometry if removed)
  * Remove that point from the array and the priority queue
  * Add the new point to the end of the array
  * Recalculate the errors of the points on either side of the removed point
* Otherwise just add the new point to the end of the sorted array

Note, however, that no attempt is made to prevent invalid geometries from forming. For example, when simplifying a polygon, the removal of points can cause lines to cross, creating invalid polygons. Any usage of this algorithm needs to consider that possibility. If the simplification is only required for visualization, and the visualization code is tolerant of line crossings, this is not an issue. But if you need valid simplified geometries, do additional validation on the final results before using them.

The concept of _simplification error_ is an estimation of the amount of damage to the Line or LinearRing that occurs when a point is removed. We have evaluated several candidates:

* cartesian triangle area (where the lat/lon coordinates are assumed to be cartesian, ie. equirectangular projection)
* triangle area (where we calculate area using Herons formula, with the triangle sides calculated using Haversin distance, to remove projection distortions)
* triangle height (triangle area divided by base length)
* cartesian height and back-paths (a very rough approximation of a Frechet distance)
* spherical height and back-paths (the same, but calculated in x/y/x coordinates)

It turns out that the above list is also ordered from fastest to slowest, and the second one, _triangle area_, gives the least visually damaging results, so is by far the best choice at this stage for general usage.

In addition, this has been developed as a separate library containing only the geometry simplification algorithm and support for simplifying:

- [x] LineString
  - [x] full geometry
  - [x] on a stream of points
- [x] Polygon
  - [x] full geometry (polygon with holes)
  - [x] on a stream of points (only outer shell)
- [x] MultiPolygon
  - [x] only full geometry (multiple shells and holes), no streaming
- [x] GeometryCollection
  - [x] only full geometries, no streaming
- [x] Non simplifiable geometries:
  - [x] Point, Circle, Rectangle

Tasks still to be done before changing this from draft:

- [x] Support Line simplification algorithms
  - [x] Triangle area
  - [ ] ~~Hausdorff distance~~
  - [x] Simple Frechet distance (no core sets)
- [x] Move to spherical coordinates (at least required in error calculation)
  - [x] Triangle area
  - [ ] ~~Hausdorff distance~~
  - [x] Simple Frechet distance (no core sets)
- [x] Extensive unit tests
  - [x] Test simplification algorithm in `GeometrySimplifierTests`
  - [x] Test error calculations in `SimplificationErrorCalculatorTests`
- [x] Write micro-benchmarks for algorithm performance
- [ ] ~~Move to Lucene PriorityQueue (and benchmark against java.util.PriorityQueue)~~
  - [ ] ~~Using Lucene `PriorityQueue.updateTop()` as possible performance optimization~~
- [x] Remove dependency on Lucene (copy SloppyMath into lib/geo)
- [x] Remove GeometrySimplfierMonitor to separate project and enhance TestMonitor for assertions
- [x] Split full-geometry simplifier and streaming simplifier into separate APIs for clarity of usage
- [ ] Write issues for proposed features:
  - [x] geo_line: https://github.com/elastic/elasticsearch/issues/87903
  - [x] TSDB geo_line: https://github.com/elastic/elasticsearch/issues/85274
  - [ ] downsampling
  - [ ] geometry simplification on ingest
  - [x] geometry simplification on query: https://github.com/elastic/elasticsearch/issues/96033
  - [ ] painless support

craigtaverner committed:
WIP line simplifier for use in geo_line

craigtaverner committed:
Generalize error calc including triangle area

The simplification error calculation can be done in several ways,
so we generalize to include triangle area calculation.

Also add more tests for both triangle area and prototype Frechet error

craigtaverner committed:
Support polygon and multipolygon simplification

craigtaverner committed:
Update docs/changelog/94859.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Refinements and more negative tests

craigtaverner committed:
Remove system.out.println

craigtaverner committed:
Export new simplification code from geo module

craigtaverner committed:
Util for selecting calculator

Used in micro-benchmarks

craigtaverner committed:
Initial microbenchmarks for geo simplification

craigtaverner committed:
Add spherical area calculation for simplification

And fix memory leak on simplification, and improve circle testing
so we can see the impact of spherical calculations more easily.

craigtaverner committed:
Fixed numerical error in flat triangle areas

Herons formula leads to NaN results when the triangle is flat, due
to small negative values inside the square root. This fix detects
those negative values and returns a zero area.

craigtaverner committed:
Add lucene license notice

craigtaverner committed:
Fixed bug with co-located points in frechet error

craigtaverner committed:
Remove dependency on lucene

And copy in SloppyMath from lucene for haversinMeters function

craigtaverner committed:
Added support for simplifier monitor

This allows code calling the GeometrySimplifier to be notified
of events related to the simplification algorithm.
For example, the project geometry-simplifier-debug uses these events to draw images and create videos of the simplification process.
And the tests for the simplifier use this to collect data to make further assertions on the results.

craigtaverner committed:
Refinements in support of visualization monitor

craigtaverner committed:
Don't create more objects than actually needed

We re-use PointError objects instead of sending them to GC.
This does require external code to re-use the simplifier itself.

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

craigtaverner committed:
Benchmark triangleheight

craigtaverner committed:
Support back-paths in simple-frechet error

craigtaverner commented on benchmarks/src/main/java/org/elasticsearch/benchmark/spatial/GeometrySimplificationBenchmark.java:
```diff
@@ -0,0 +1,94 @@
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@State(Scope.Thread)
+public class GeometrySimplificationBenchmark {
+    @Param({ "cartesiantrianglearea", "triangleArea", "frechetError" })
```
Fixed in 65cf7067634331db4ae88d2ec50fd99ca9947294

craigtaverner commented on libs/geo/src/main/java/org/elasticsearch/geometry/simplify/SimplificationErrorCalculator.java:
```diff
@@ -0,0 +1,168 @@
+                double rightYrotated = rightY * cos - rightX * sin;
+                assert Math.abs(rightYrotated) < 1e-10;
+                assert Math.abs(rightXrotated - len) < len / 1e10;
+                // Return distance to x-axis TODO: also include back-paths for Frechet distance calculation
```
Fixed in a0b5e05ad790761983dde3fa67e034767772a562

craigtaverner commented:
After implementing back-paths in the Frechet distance calculation, I re-ran benchmarks:

```
Benchmark                                      (calculatorName)  (maxPoints)  Mode  Cnt  Score   Error  Units
GeometrySimplificationBenchmark.simplify  cartesiantrianglearea         1000  avgt    5  2,756 ± 0,063  ms/op
GeometrySimplificationBenchmark.simplify           triangleArea         1000  avgt    5  5,438 ± 0,092  ms/op
GeometrySimplificationBenchmark.simplify         triangleheight         1000  avgt    5  5,722 ± 0,134  ms/op
GeometrySimplificationBenchmark.simplify     simplefrechetError         1000  avgt    5  3,181 ± 0,066  ms/op
```

So it looks pretty good, however, this is still Cartesian. If I port to Spherical, it will get slower. So at this point `trianglearea` is still the primary one to consider, despite being a bit slow.

iverase commented on libs/geo/src/main/java/org/elasticsearch/geometry/simplify/GeometrySimplifier.java:
```diff
@@ -0,0 +1,482 @@
+    /**
+     * Consume a single point on the stream of points to be simplified
+     */
+    public void consume(double x, double y) {
```
Make it protected and final?

iverase commented on libs/geo/src/main/java/org/elasticsearch/geometry/simplify/GeometrySimplifier.java:
```diff
@@ -0,0 +1,482 @@
+    /**
+     * Simplify an entire geometry in a non-streaming fashion
+     */
+    public abstract T simplify(T geometry);
```
I think this is wrong as we are still using the same streaming algorithm. This is just a utility method to perform the streaming algorithm on a geometry?

craigtaverner replied:
OK. What I did from the beginning is target two use cases, streaming and non-streaming. I also have a bunch of tests asserting that only very limited geometry types can access the streaming case (simple LineString, Polygon with no holes), while all other geometry types can use the non-streaming API. This was because I perceived the non-streaming use cases as very valuable, even if not the first use case to be implemented.

Also, I use the non-streaming API for most tests, and benchmarks as well as the visualisation tool (images and animations used in the presentation I gave).

iverase reviewed (commented):
I think the approach looks good. One thing that is confusing is that at the moment the API looks like a common geometry simplifier but it is important to note that the difference with other approaches is that we are doing the simplification on a streaming fashion, do let's try to make it more obvious for the users?

craigtaverner committed:
Added GeometryCollection support

And pass simplifier start/end events even for non simplified
geometries, for improved symmetry.

craigtaverner committed:
Generalize Point simplifier to identity simplifier

So we can complete support for all geometry types.

craigtaverner committed:
Use identity simplifier for Rectangle, Circle and MultiPoint

craigtaverner commented on libs/geo/src/main/java/org/elasticsearch/geometry/simplify/SimplificationErrorCalculator.java:
```diff
@@ -0,0 +1,189 @@
+import java.util.List;
+import java.util.Locale;
+
+public interface SimplificationErrorCalculator {
```
The registry is used by the benchmark, which needs to be parameterized by String[], and then lookup the actual calculator by name. An earlier version used a switch statement in the byName method, but I thought the registry looked a bit more elegant.

While I could move this all into the benchmark code instead, I did plan to have the tests make use of this also. Right now the tests require extending the GeometrySimplifierTests class with calculator specific sub-classes, that is not very elegant, and I would have preferred to use test parameterization, like we do in the benchmarks. I'm being blocked by ESTestCase which seems to dislike parameterized tests.

craigtaverner replied:
On a separate note, the reason I did not use lambda syntax was that to support the actual paper you originally provided, we need to maintain a little more state in the calculators, and they would either need constructors and fields, or additional methods for this. Right now I do not have any calculators that need that, and renamed my FrechetError to SimpleFrechetError to emphasize that I'm not doing the full calculation, only a simplified version. But I did imagine that we might take that step (I even expected that before finishing this work, but no longer think so).

craigtaverner replied:
I did investigate this a little further, and only one of the calculators contained a single function so could easily be converted to lambda syntax, so I did not take that step. However, I did make constants and removed the Registry class, which hopefully covers most of what you wanted.

craigtaverner commented:
So I'm thinking more about the fact that this API currently does both streaming and non-streaming, and that most tests, benchmarks and the visualisation library use the non-streaming API, which internally just delegates to the streaming API.

However, I can imagine one improvement that should satisfy your interest in making the two use cases clearer and more separated, and that would be to split them into two interfaces, and the streaming one only has two implementations, LineString and Polygon, and the non-streaming one could continue to use instances of the streaming objects internally, but not expose their APIs. This would be a little work, but quite achievable, would more clearly separate the two, and also get rid of the current code throwing errors like:

    throw new IllegalArgumentException("MultiPolygon geometry simplifier cannot work in streaming mode");

Getting rid of those lines, and therefor also removing any runtime checks for usage of the two APIs would certainly be an improvement.

iverase commented:
> I'm thinking more about the fact that this API currently does both streaming and non-streaming,
I don't think that is right. The code is using an streaming approach for line simplification but it accepts as an input a geometry.

craigtaverner committed:
Split out streaming simplifier for clarity

The original GeometrySimplifier provided both a streaming and
non-streaming interface. The streaming interface only works for
limited geometries, necessitating runtime checks and throwing
exceptions to ensure correct usage. This was potentially confusing
and error-prone. So now we've split the two classes:

* StreamingGeometrySimplifier - only supports streaming mode, with methods consume(x, y) and produce() and works only with four geometry types: Point, Line, LinearRing, Polygon (with no holes)
* GeometrySimplifier - only supports non-streaming mode with the method simplify(geometry) and works with almost all geometry types

craigtaverner commented on libs/geo/src/main/java/org/elasticsearch/geometry/simplify/SimplificationErrorCalculator.java:
```diff
@@ -0,0 +1,189 @@
+     * a good enough approximation of the original line. This restriction is currently true of all the
+     * calculations implemented so far.
+     */
+    class SimpleFrechetErrorCalculator implements SimplificationErrorCalculator {
```
While this is based on the FrechetError oracle from the Abam et al. paper, it is not the complete calculation, since it only considers the currently simplified geometry, not the impact on the original geometry. The prefix `Simple` is probably not sufficient to cover this difference, so a better name would be `HeightAndBackpathDistanceCalculator`, which describes literally what it does.

craigtaverner committed:
Renamed Frechet error and removed registry

To simplify the code a little, we removed the calculator registry and also renamed the frechet error to describe what it actually does, instead of implying it was an actual Frechet distance, which it is only an approximation of.

craigtaverner committed:
Improved comments

craigtaverner committed:
WIP Added spherical height and backpath calculator

craigtaverner committed:
Test for spherical backpath calculator

craigtaverner commented:
I got a basic version of a spherical version of the simplified frechet calculation, and it is much slower:

```
Benchmark                                                   (calculatorName)  (maxPoints)  Mode  Cnt   Score   Error  Units
GeometrySimplificationBenchmark.simplify               cartesiantrianglearea         1000  avgt    5   2,745 ± 0,096  ms/op
GeometrySimplificationBenchmark.simplify                        triangleArea         1000  avgt    5   5,398 ± 0,035  ms/op
GeometrySimplificationBenchmark.simplify                      triangleheight         1000  avgt    5   5,565 ± 0,010  ms/op
GeometrySimplificationBenchmark.simplify           heightandbackpathdistance         1000  avgt    5   3,074 ± 0,006  ms/op
GeometrySimplificationBenchmark.simplify  sphericalheightandbackpathdistance         1000  avgt    5  15,128 ± 0,185  ms/op
```

So the current winner balancing performance with accuracy, is still `trianglearea` (spherical coords).

craigtaverner committed:
spotless

craigtaverner committed:
Javadocs warning of invalid polygons

Note, however, that no attempt is made to prevent invalid geometries from forming. For example, when simplifying a polygon,
the removal of points can cause lines to cross, creating invalid polygons. Any usage of this algorithm needs to consider that
possibility. If the simplification is only required for visualization, and the visualization code is tolerant of line crossings,
this is not an issue. But if you need valid simplified geometries, do additional validation on the final results before using them.

iverase reviewed (approved):
LGTM

Thank you for all the javadocs, now it makes much more sense to me
